src/wh_server.c

```diff
@@ -119,35 +728,1492 @@ static int _wh_Server_HandleNvmRequest(whServerContext* server,
+            sig = (uint8_t*)(&packet->pkEccVerifyReq + 1);
+            hash = (uint8_t*)(&packet->pkEccVerifyReq + 1) +
+                packet->pkEccVerifyReq.sigSz;
+            /* init public key */
+            ret = wc_ecc_init_ex(eccPublic, ctx->heap, ctx->devId);
+            /* load the public key */
+            if (ret == 0) {
+                meta->id = packet->pkEccVerifyReq.keyId;
+                ret = hsmLoadKeyEcc(ctx, meta, eccPublic,
+                    packet->pkEccVerifyReq.curveId);
+            }
+            /* verify the signature */
+            if (ret == 0) {
+                ret = wc_ecc_verify_hash(sig, packet->pkEccVerifyReq.sigSz,
+                    hash, packet->pkEccVerifyReq.hashSz, (int*)&field,
+                    eccPublic);
+            }
+            wc_ecc_free(eccPublic);
+            if (ret == 0) {
+                packet->pkEccVerifyRes.res = field;
+                /* set len */
+                packet->len = sizeof(packet->pkEccVerifyRes);
+                /* set type */
+                packet->subType = WC_ALGO_TYPE_PK;
+            }
+            break;
+        case WC_PK_TYPE_EC_CHECK_PRIV_KEY:
+            /* init private key */
+            ret = wc_ecc_init_ex(crypto.eccPrivate, ctx->heap, ctx->devId);
+            /* load the private key */
+            if (ret == 0) {
+                meta->id = packet->pkEccCheckReq.keyId;
+                ret = hsmLoadKeyEcc(ctx, meta, crypto.eccPrivate,
+                    packet->pkEccCheckReq.curveId);
+            }
+            /* check the key */
+            if (ret == 0) {
+                ret = wc_ecc_check_key(crypto.eccPrivate);
+            }
+            wc_ecc_free(crypto.eccPrivate);
+            if (ret == 0) {
+                packet->pkEccCheckRes.ok = 1;
+                /* set len */
+                packet->len = sizeof(packet->pkEccCheckRes);
+                /* set type */
+                packet->subType = WC_ALGO_TYPE_PK;
+            }
+            break;
+#endif
+        case WC_PK_TYPE_CURVE25519_KEYGEN:
+            /* init private key */
+            ret = wc_curve25519_init_ex(server->crypto->curve25519Private, NULL,
+                INVALID_DEVID);
+            /* make the key */
+            if (ret == 0) {
+                ret = wc_curve25519_make_key(server->crypto->rng,
+                    packet->pkCurve25519kgReq.sz,
+                    server->crypto->curve25519Private);
+            }
+            /* cache the generated key */
+            if (ret == 0) {
+                ret = hsmCacheKeyCurve25519(server, server->crypto->curve25519Private);
+            }
+            /* set the assigned id */
+            wc_curve25519_free(server->crypto->curve25519Private);
+            if (ret > 0) {
+                packet->pkCurve25519kgRes.keyId = ret;
+                *size = sizeof(packet->pkCurve25519kgRes);
+            }
+            else
+                ret = BAD_FUNC_ARG;
+            break;
```

Comment:
on success this will return the key ID not a valid return code. We need to refactor hsmCacheKey to return a proper error code, or at the very least, fix all cases of mixed up return code/key IDs in this function